src/Compiler/Opts/ConstantFold.idr

```diff
@@ -0,0 +1,164 @@
+    inlineLetEnv out val (CRef fc y) = CRef fc y
+    inlineLetEnv out val (CLam fc y z) = CLam fc y $ inlineLetEnv (suc out) val z
+    inlineLetEnv out val (CLet fc y inlineOK z w) = CLet fc y inlineOK (inlineLetEnv out val z) (inlineLetEnv (suc out) val w)
+    inlineLetEnv out val (CApp fc y xs) = CApp fc (inlineLetEnv out val y) (inlineLetEnv out val <$> xs)
+    inlineLetEnv out val (CCon fc y z tag xs) = CCon fc y z tag $ inlineLetEnv out val <$> xs
+    inlineLetEnv out val (COp fc y xs) = COp fc y $ inlineLetEnv out val <$> xs
+    inlineLetEnv out val (CExtPrim fc p xs) = CExtPrim fc p $ inlineLetEnv out val <$> xs
+    inlineLetEnv out val (CForce fc y z) = CForce fc y $ inlineLetEnv out val z
+    inlineLetEnv out val (CDelay fc y z) = CDelay fc y $ inlineLetEnv out val z
+    inlineLetEnv out val (CConCase fc sc xs y) = CConCase fc (inlineLetEnv out val sc) (inlineLetConAlt out val <$> xs) (inlineLetEnv out val <$> y)
+    inlineLetEnv out val (CConstCase fc sc xs y) = CConstCase fc (inlineLetEnv out val sc) (inlineLetConstAlt out val <$> xs) (inlineLetEnv out val <$> y)
+    inlineLetEnv out val (CPrimVal fc y) = CPrimVal fc y
+    inlineLetEnv out val (CErased fc) = CErased fc
+    inlineLetEnv out val (CCrash fc y) = CCrash fc y
+
+    inlineLetConAlt : SizeOf outer -> CExp inner -> CConAlt (outer ++ [x] ++ inner) -> CConAlt (outer ++ inner)
+    inlineLetConAlt out val (MkConAlt y z tag args w) = MkConAlt y z tag args $
+        replace {p=CExp} (sym $ appendAssociative args outer inner) $
+        inlineLetEnv (mkSizeOf args + out) val $
+        replace {p=CExp} (appendAssociative args outer ([x] ++ inner)) w
+
+    inlineLetConstAlt : SizeOf outer -> CExp inner -> CConstAlt (outer ++ [x] ++ inner) -> CConstAlt (outer ++ inner)
+    inlineLetConstAlt out val (MkConstAlt c e) = MkConstAlt c $ inlineLetEnv out val e
+
+  export
+  inlineLet : CExp vars -> CExp (x :: vars) -> CExp vars
+  inlineLet val exp = inlineLetEnv zero val exp
+
+findConstAlt : Constant -> List (CConstAlt vars) -> Maybe (CExp vars) -> Maybe (CExp vars)
+findConstAlt c [] def = def
+findConstAlt c (MkConstAlt c' exp :: alts) def = if c == c'
+    then Just exp
+    else findConstAlt c alts def
+
+foldableOp : PrimFn ar -> Bool
+foldableOp BelieveMe = False
+foldableOp (Cast from to) = fromMaybe False [| intKind from `smaller` intKind to |]
+  where
+    smaller : IntKind -> IntKind -> Bool
+    smaller (Signed x) (Signed y) = x <= y
+    smaller (Unsigned x) (Unsigned y) = x <= y
+    smaller (Signed x) (Unsigned y) = x < P y
+    smaller (Unsigned x) (Signed y) = P x < y
+foldableOp _ = True
+
+-- constant folding of primitive functions
+-- if a primitive function is applied to only constant
+-- then replace with the result
+-- if there's only 1 constant argument to a commutative function
+-- move the constant to the right. This simplifies Compiler.Opts.Identity
+constFold : {vars : _} -> CExp vars -> CExp vars
```

Comment:
I wonder whether this could have a more general type so that you're not
forced to call `inlineLet` before re-traversing the term (and therefore
the inlined term too!).